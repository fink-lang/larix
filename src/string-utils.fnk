unindent_text = fn text, ind:
  [first, ...rest] = text.split rx/\n/g

  [...lines] = pipe rest:
    map line:
      line.slice ind

  [first, ...lines].join '\n'
