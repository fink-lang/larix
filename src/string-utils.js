import {curr_loc, next_loc} from '@fink/prattler';


export const unindent_text = (text, ind)=> (
  text
    .split(/\n/g)
    .map((line, idx)=> line.slice(idx > 0 ? ind : 0))
    .join('\n')
);


export const strip_block = (strings, ...parts)=> {
  const [, ...lines] = String.raw({raw: strings}, ...parts).split('\n');
  const ind = lines[0].search(/[^ ]/);
  const str = lines.map((lne)=> lne.slice(ind)).join('\n');
  return str;
};

