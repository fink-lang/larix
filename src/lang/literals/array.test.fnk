{describe, it, expect, to_match_snapshot} = import '@fink/jest'

{parse_expr, parse} = import '../../'


describe 'array [...]', fn:

  it 'parses empty: []', fn:
    expect
      parse_expr `[]`
      to_match_snapshot


  it 'parses single elemement: [1]', fn:
    expect
      parse_expr `[1]`
      to_match_snapshot


  it 'parses multiple elements: [1, 2]', fn:
    expect
      parse_expr `[1, 2]`
      to_match_snapshot

    expect
      parse_expr `[1, 2, 3]`
      to_match_snapshot


  it 'parses leading commas: [,, foo]', fn:
    expect
      parse_expr `[,, foo]`
      to_match_snapshot


  it 'parses middle commas: [,, foo]', fn:
    expect
      # [a, , ..., b, c]
      parse_expr `[foo, ,bar]`
      to_match_snapshot


  it 'parses trailing commas: [...foo,,]', fn:
    expect
      parse_expr `[...foo, ]`
      to_match_snapshot

    expect
      parse_expr `[...foo, , ]`
      to_match_snapshot

    expect
      parse_expr `[, ...foo, ]`
      to_match_snapshot


describe 'array block', fn:
  it 'parses empty', fn:
    expect
      parse_expr `
        [
        ]`
      to_match_snapshot


  it 'parses single elemement', fn:
    expect
      parse_expr `
        [
          1
        ]`
      to_match_snapshot


  it 'parses multiple elements', fn:
    expect
      parse_expr `
        [
          1
          2
        ]`
      to_match_snapshot

    expect
      parse_expr `
        [
          1, 2
        ]`
      to_match_snapshot


  it 'parses leading commas', fn:
    expect
      parse_expr `
        [
          ,
          ,
          foo
        ]`
      to_match_snapshot

    expect
      parse_expr `
        [
          ,,
          foo
        ]`
      to_match_snapshot


  it 'parses trailing commas', fn:
    expect
      parse_expr `
        [
          foo
          ,
        ]`
      to_match_snapshot

    expect
      parse_expr `
        [
          foo
          ,
          ,
        ]`
      to_match_snapshot

    expect
      parse_expr `
        [
          ,
          ...foo
          ,
        ]`
      to_match_snapshot


describe 'list', fn:
  it 'parses single elemement', fn:
    expect
      parse_expr `
        list:
          1
        `
      to_match_snapshot


  it 'parses multiple elements', fn:
    expect
      parse_expr `
        list:
          1
          # comment
          2
        `
      to_match_snapshot


  it 'parses leading commas', fn:
    expect
      parse_expr `
        list:
          ,
          ,
          foo
        `
      to_match_snapshot

    expect
      parse_expr `
        list:
          ,,
          foo
        `
      to_match_snapshot


  it 'parses trailing commas', fn:
    expect
      parse_expr `
        list:
          foo
          ,
        `
      to_match_snapshot

    expect
      parse_expr `
        list:
          foo
          ,
          ,
        `
      to_match_snapshot

    expect
      parse_expr `
        list:
          ,
          ...foo
          ,
        `
      to_match_snapshot


describe 'array [...] - parsing failures', fn:
  it `errors when missing ']'`, fn:
    expect
      parse `[`, 'test.fnk'
      to_match_snapshot


  it `errors when missing ','`, fn:
    expect
      parse `[1)]`, 'test.fnk'
      to_match_snapshot


