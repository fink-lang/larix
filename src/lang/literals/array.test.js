import {parse_expr} from '../../';
import {strip_block} from '../../string-utils';


describe('array [...]', ()=> {

  it('parses empty: []', ()=> {
    expect(
      parse_expr(`[]`)
    ).toMatchSnapshot();
  });


  it('parses single elemement: [1]', ()=> {
    expect(
      parse_expr(`[1]`)
    ).toMatchSnapshot();
  });


  it('parses multiple elements: [1, 2, 3]', ()=> {
    expect(
      parse_expr(`[1, 2, 3]`)
    ).toMatchSnapshot();
  });


  it('parses leading commas: [,, foo]', ()=> {
    expect(
      parse_expr(`[,, foo]`)
    ).toMatchSnapshot();
  });
});


describe('array [...] - parsing failures', ()=> {
  it('throws when missing `]`', ()=> {
    expect(
      ()=> parse_expr(`[`)
    ).toThrow(strip_block`
      Unexpected end of code:
      1| [
         ^`
    );
  });

  it('throws when missing `,`', ()=> {
    expect(
      ()=> parse_expr(`[1)]`)
    ).toThrow(strip_block`
      Expected ']' but found ')':
      1| [1)]
           ^`
    );
  });

  it('foo', ()=> {
    expect(
      ()=> parse_expr(`[] []`)
    ).toMatchSnapshot();
  });
});

