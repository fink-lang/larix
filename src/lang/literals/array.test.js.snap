// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array [...] - parsing failures foo 1`] = `[Function]`;

exports[`array [...] parses empty: [] 1`] = `
array (1:0-1:2)

`;

exports[`array [...] parses leading commas: [,, foo] 1`] = `
array (1:0-1:8)


  ident (1:4-1:7) foo
`;

exports[`array [...] parses multiple elements: [1, 2, 3] 1`] = `
array (1:0-1:9)
  number (1:1-1:2) 1
  number (1:4-1:5) 2
  number (1:7-1:8) 3
`;

exports[`array [...] parses single elemement: [1] 1`] = `
array (1:0-1:3)
  number (1:1-1:2) 1
`;
